qmail-autoresponder-mysql.1
.\" $Id$
.TH qmail-autoresponder-mysql 1
.SH NAME
qmail-autoresponder-mysql \- Produce automatic responses with qmail
.SH SYNOPSIS
.B qmail-autoresponder-mysql
[
.B \-cLNqT
] [
.BR "-O NAME" [ =VALUE ]
] [
.B -n NUM
] [
.B -s STR
] [
.B -t TIME
]
.I USERNAME DOMAIN
[
.I TABLE_PREFIX
]
.SH DESCRIPTION
.B qmail-autoresponder-mysql
sends an automatically-generated message in response to an incoming
message.  By default, it limits the number of responses it sends to a
maximum of one message per hour per sender.  It will not respond to
any known mailing list, nor will it respond to either bounces or
loops.
.P
The
.I USERNAME
and
.I DOMAIN
arguments are used to select which autoresponder to use from the MySQL
database.  If a record in the
.B autoresponder
table matches the username and domain, the program purges old records
for that autoresponder from the
.B response
table and then counts the remaining responses listed there.  A record is
then added to that table indicating if a response was sent or not.  If a
table named
.B log
exists, the same record is added to it as well.  The
.B log
table is never automatically pruned.
.P
If a
.I TABLE_PREFIX
argument is given, all table names are prefixed with that string.  For
example, if
.I TABLE_PREFIX
is
.B q_
then the tables used are
.B q_autoresponder
and
.BR q_response .
This can be used to allow these tables to co-exist with tables for other
programs in the same database.
.P
See the man page for
.BR qmail-autoresponder (1)
for more details on options and general autoresponder behavior.
.SH OPTIONS
.B qmail-autoresponder-mysql
supports options specified on the command line and as fields in the
.B autoresponder
table.  Option fields are named
.B opt_\fIOPTION
in the database, and override options set on the command line.  See the
man page for
.BR qmail-autoresponder (1)
for a full list of options.
.SH EXAMPLES
Typical usage in a
.B .qmail-default
file would be as follows:

.EX
|qmail-autoresponder-mysql -- "$EXT" "$HOST"
.EE
.SH ENVIRONMENT
.TP
.B DTLINE
.B qmail-autoresponder-mysql
searches for this line (generated by qmail-local) in the input and
bounces the message if it is present (which indicates a mail loop).
It then adds the line to the output.
.TP
.B MYSQL_DB
.TP
.B MYSQL_HOST
.TP
.B MYSQL_PASS
.TP
.B MYSQL_SOCKET
.TP
.B MYSQL_USER
These set the default parameters for the MySQL database connection.  If
not set,
.B qmail-autoresponder-mysql
uses the defaults in the
.B qmail-autoresponder
group in the MySQL option file.
.TP
.B SENDER
This contains the envelope sender address and is used as the address
to send the reply to.
.SH "SEE ALSO"
qmail-autoresonder(1)


NEWS
-------------------------------------------------------------------------------
Changes in version 0.98

- Manually invoke qmail-inject and qmail-queue separately in order to
  fully control the envelope and capture the qmail PID.

- Added a "bcc" option to allow for copying autoresponses to an address
  other than the sender.

- Add an optional separator before the copied message.

- Add a newline after the response message and after the copied header,
  to make sure there is at least one blank line between them.

- The MySQL autoresponder now supports an optional table name prefix to
  better interoperate with other applications using the same database.

- Added support for a log file or table that is not pruned.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.97

- Completely rewrote header handling, improving its handling of headers
  split across multiple lines among other things.

- Added support for omitting parts of the copied headers.

- Internal MIME headers are now stripped from multipart messages.

- Added support for limiting the number of lines copied from the
  original message body.

- Added support for setting autoresponder options through files.

- Added an option to control the generation of the In-Reply-To: header.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.96.2

- Fixed up the multi-line strings that no longer compile with recent
  compilers.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.96.1

- Added missing MySQL schema to package.

- Switched to new bglibs (1.006) library usage.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.96

- Added code to use MySQL instead of files for all the autoresponse
  data.

- Removed the option to omit the "To:" header, to allow for copying the
  response to extra addresses by adding BCCs to the message.

- Switched to using external bglibs.

- Removed adding the Delivered-To: line to outgoing messages.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.95

- Made the "MESSAGE-FILE" parameter optional, defaulting to
  "message.txt" in the autoresponder directory.

- Added a patch from Erik Sjoelund <erik@xpedio.com> to insert the
  original message's message-ID into a new In-Reply-To header field.

- Ignore messages containing the header fields listed in RFC 2369:
	List-(Help|Unsubscribe|Subscribe|Post|Owner|Archive):

- Portability fixes and documentation bugs.

- Added a manual page.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.93

- Fixed an off-by-one bug in the rate limiting logic (again).

- Fixed an off-by-one bug in the Delivered-To checking code.

- The output writing routine will now substitute "%S" in the
  autoresponse message with the original message's subject.
-------------------------------------------------------------------------------
Changes in version 0.92

- Fixed the rate limiting logic that would prevent messages from being
  sent out at the appropriate interval.

- Fixed bug in option parsing.
-------------------------------------------------------------------------------
Changes in version 0.91

- Added an option to insert the original subject into the reply.

- Simplified the header scanning logic.
-------------------------------------------------------------------------------


qmail-autoresponder-mysql.c
#include <mysql/mysql.h>
#include <stdlib.h>
#include <string.h>
#include <str/str.h>
#include "qmail-autoresponder.h"

const char usage_args[] = "USERNAME DOMAIN [TABLE_PREFIX]";
const char usage_post[] =
"  All the options except for -q, -D, and -N can be overridden\n"
"  in the autoresponder SQL table.\n";

static long autoresponder;
static MYSQL mysql;
static str query;
static const char* prefix;

static int str_cats_quoted(str* s, const char* p)
{
  if (!str_catc(s, '\'')) return 0;
  for (; *p != 0; ++p) {
    switch (*p) {
    case '\'':
    case '\\':
      if (!str_catc(s, '\\')) return 0;
    default:
      if (!str_catc(s, *p)) return 0;
    }
  }
  return str_catc(s, '\'');
}

static MYSQL_RES* do_select(const char* username, const char* domain)
{
  MYSQL_RES* result;
  str_copy3s(&query,
	     "SELECT * "
	     "FROM ", prefix, "autoresponder "
	     "WHERE username");
  if (username) {
    str_catc(&query, '=');
    str_cats_quoted(&query, username);
  }
  else
    str_cats(&query, " IS NULL");
  str_cats(&query, " AND domain=");
  str_cats_quoted(&query, domain);
  if (mysql_real_query(&mysql, query.s, query.len))
    fail_temp("Could not select autoresponder.");
  if ((result = mysql_store_result(&mysql)) == 0)
    fail_temp("Error fetching result from database.");
  return result;
}

static void handle_field(const char* name,
			 const char* value,
			 unsigned int length)
{
  if (strcmp(name, "id") == 0)
    autoresponder = atol(value);
  else if (strcmp(name, "response") == 0)
    str_copyb(&response, value, length);
  else if (memcmp(name, "opt_", 4) == 0)
    handle_option(name + 4, value, length);
}

void init_autoresponder(int argc, char** argv)
{
  const char* username;
  const char* domain;
  MYSQL_RES* result;
  MYSQL_ROW row;
  unsigned long* lengths;
  unsigned int num_fields;
  unsigned int i;
  MYSQL_FIELD* fields;
  
  if (argc < 2 || argc > 3)
    usage("Incorrect number of command-line arguments.");
  username = argv[0];
  domain = argv[1];
  prefix = (argc > 2) ? argv[2] : "";

  mysql_init(&mysql);
  mysql_options(&mysql, MYSQL_READ_DEFAULT_GROUP, "qmail-autoresponder");
  if (!mysql_real_connect(&mysql, getenv("MYSQL_HOST"),
			  getenv("MYSQL_USER"), getenv("MYSQL_PASS"),
			  getenv("MYSQL_DB"),
			  0, getenv("MYSQL_SOCKET"), 0))
    fail_temp("Could not connect to MySQL");

  result = do_select(username, domain);
  if (mysql_num_rows(result) == 0) {
    mysql_free_result(result);
    result = do_select(NULL, domain);
    if (mysql_num_rows(result) == 0)
      exit(0);
  }

  if ((num_fields = mysql_num_fields(result)) == 0
      || (fields = mysql_fetch_fields(result)) == 0
      || (row = mysql_fetch_row(result)) == 0
      || (lengths = mysql_fetch_lengths(result)) == 0)
    fail_temp("Error fetching autoresponder from database.");
  else {
    for (i = 0; i < num_fields; ++i)
      if (row[i])
	handle_field(fields[i].name, row[i], lengths[i]);
    mysql_free_result(result);
  }
}

int count_history(const char* sender)
{
  unsigned count = opt_msglimit;
  int send_response;
  MYSQL_RES* result;
  MYSQL_ROW row;
  
  if (!opt_nodelete) {
    str_copy3s(&query, "DELETE FROM ", prefix, "response "
	       "WHERE autoresponder=");
    str_cati(&query, autoresponder);
    str_cats(&query, " AND timestamp < (now() - INTERVAL ");
    str_catu(&query, opt_timelimit);
    str_cats(&query, " SECOND)");
    if (mysql_real_query(&mysql, query.s, query.len))
      fail_temp("Could not delete old records from database.");
  }
  
  str_copy3s(&query, "SELECT count(*) "
	     "FROM ", prefix, "response "
	     "WHERE sent_response <> 0 "
	     "AND autoresponder=");
  str_cati(&query, autoresponder);
  str_cats(&query, " AND sender=");
  str_cats_quoted(&query, sender);
  str_cats(&query, " AND timestamp > (now() - INTERVAL ");
  str_catu(&query, opt_timelimit);
  str_cats(&query, " SECOND)");
  if (mysql_real_query(&mysql, query.s, query.len))
    fail_temp("Could not select count of records from database.");
  if ((result = mysql_store_result(&mysql)) == 0 ||
      (row = mysql_fetch_row(result)) == 0)
    fail_temp("Error fetching count of records from database.");
  else {
    count = atol(row[0]);
    mysql_free_result(result);
  }

  send_response = count < opt_msglimit;
  
  return send_response;
}

static int logit(const char* table, const char* sender, int responded)
{
  str_copy4s(&query,
	     "INSERT INTO ", prefix, table, " "
	     "(autoresponder,timestamp,sent_response,sender) "
	     "VALUES (");
  str_cati(&query, autoresponder);
  str_cats(&query, ",now(),");
  str_catu(&query, responded);
  str_catc(&query, ',');
  str_cats_quoted(&query, sender);
  str_cats(&query, ")");
  return mysql_real_query(&mysql, query.s, query.len);
}

void log_sender(const char* sender, int responded)
{
  if (logit("response", sender, responded))
    fail_temp("Could not insert response record into database.");
  logit("log", sender, responded); /* Ignore errors for the log table */
}
